Mathlib/Data/Set/Basic.lean

```diff
@@ -2754,3 +2754,204 @@ instance decidableSetOf (p : α → Prop) [Decidable (p a)] : Decidable (a ∈ {
+  Set.disjoint_iff.not.trans <| not_forall.trans <| exists_congr fun _ => not_not
+#align set.not_disjoint_iff Set.not_disjoint_iff
+
+theorem not_disjoint_iff_nonempty_inter : ¬Disjoint s t ↔ (s ∩ t).Nonempty :=
+  not_disjoint_iff
+#align set.not_disjoint_iff_nonempty_inter Set.not_disjoint_iff_nonempty_inter
+
+alias not_disjoint_iff_nonempty_inter ↔ _ Nonempty.not_disjoint
+
+theorem disjoint_or_nonempty_inter (s t : Set α) : Disjoint s t ∨ (s ∩ t).Nonempty :=
+  (em _).imp_right not_disjoint_iff_nonempty_inter.mp
+#align set.disjoint_or_nonempty_inter Set.disjoint_or_nonempty_inter
+
+theorem disjoint_iff_forall_ne : Disjoint s t ↔ ∀ x ∈ s, ∀ y ∈ t, x ≠ y := by
+  simp only [Ne.def, disjoint_left, @imp_not_comm _ (_ = _), forall_eq']
+#align set.disjoint_iff_forall_ne Set.disjoint_iff_forall_ne
+
+theorem _root_.Disjoint.ne_of_mem (h : Disjoint s t) {x y} (hx : x ∈ s) (hy : y ∈ t) : x ≠ y :=
+  disjoint_iff_forall_ne.mp h x hx y hy
+#align disjoint.ne_of_mem Disjoint.ne_of_mem
+
+theorem disjoint_of_subset_left (h : s ⊆ u) (d : Disjoint u t) : Disjoint s t :=
+  d.mono_left h
+#align set.disjoint_of_subset_left Set.disjoint_of_subset_left
+
+theorem disjoint_of_subset_right (h : t ⊆ u) (d : Disjoint s u) : Disjoint s t :=
+  d.mono_right h
+#align set.disjoint_of_subset_right Set.disjoint_of_subset_right
+
+theorem disjoint_of_subset {s t u v : Set α} (h1 : s ⊆ u) (h2 : t ⊆ v) (d : Disjoint u v) :
+    Disjoint s t :=
+  d.mono h1 h2
+#align set.disjoint_of_subset Set.disjoint_of_subset
+
+@[simp]
+theorem disjoint_union_left : Disjoint (s ∪ t) u ↔ Disjoint s u ∧ Disjoint t u :=
+  disjoint_sup_left
+#align set.disjoint_union_left Set.disjoint_union_left
+
+@[simp]
+theorem disjoint_union_right : Disjoint s (t ∪ u) ↔ Disjoint s t ∧ Disjoint s u :=
+  disjoint_sup_right
+#align set.disjoint_union_right Set.disjoint_union_right
+
+theorem disjoint_diff {a b : Set α} : Disjoint a (b \ a) :=
+  disjoint_iff.2 (inter_diff_self _ _)
+#align set.disjoint_diff Set.disjoint_diff
+
+@[simp]
+theorem disjoint_empty (s : Set α) : Disjoint s ∅ :=
+  disjoint_bot_right
+#align set.disjoint_empty Set.disjoint_empty
+
+@[simp]
+theorem empty_disjoint (s : Set α) : Disjoint ∅ s :=
+  disjoint_bot_left
+#align set.empty_disjoint Set.empty_disjoint
+
+@[simp]
+theorem univ_disjoint {s : Set α} : Disjoint univ s ↔ s = ∅ :=
+  top_disjoint
+#align set.univ_disjoint Set.univ_disjoint
+
+@[simp]
+theorem disjoint_univ {s : Set α} : Disjoint s univ ↔ s = ∅ :=
+  disjoint_top
+#align set.disjoint_univ Set.disjoint_univ
+
+@[simp]
+theorem disjoint_singleton_left {a : α} {s : Set α} : Disjoint {a} s ↔ a ∉ s :=
+  by simp [Set.disjoint_iff, subset_def]
+#align set.disjoint_singleton_left Set.disjoint_singleton_left
+
+@[simp]
+theorem disjoint_singleton_right {a : α} {s : Set α} : Disjoint s {a} ↔ a ∉ s :=
+  by rw [Disjoint.comm]; exact disjoint_singleton_left
+#align set.disjoint_singleton_right Set.disjoint_singleton_right
+
+@[simp 1100]
```

Comment:
The best solution is probably to not make it a simp-lemma at all.
```suggestion
```

Reply:
I suspect this will need to fix more proofs. I found `disjoint_singleton` can be proved by `simp` if making `disjoint_singleton_left` have a greater priority than `disjoint_singleton_right`. I'm not sure if it's better.